lib/template-loader.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Block template loader functions.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Adds necessary filters to use 'wp_template' posts instead of theme template files.
+ */
+function gutenberg_add_template_loader_filters() {
+	if ( ! post_type_exists( 'wp_template' ) ) {
+		return;
+	}
+
+	/**
+	 * Array of all overrideable default template types.
+	 *
+	 * @see get_query_template
+	 *
+	 * @var array
+	 */
+	$template_types = array(
+		'index',
+		'404',
+		'archive',
+		'author',
+		'category',
+		'tag',
+		'taxonomy',
+		'date',
+		// Skip 'embed' for now because it is not a regular template type.
+		'home',
+		'frontpage',
+		'privacypolicy',
+		'page',
+		'search',
+		'single',
+		'singular',
+		'attachment',
+	);
+	foreach ( $template_types as $template_type ) {
+		add_filter( $template_type . '_template', 'gutenberg_override_query_template', 20, 3 );
+	}
+
+	add_filter( 'template_include', 'gutenberg_find_template', 20 );
+}
+add_action( 'wp_loaded', 'gutenberg_add_template_loader_filters' );
+
+/**
+ * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ *
+ * The method returns an empty result for every template so that a 'wp_template' post
+ * is used instead.
+ *
+ * @see gutenberg_find_template
+ *
+ * @param string $template  Path to the template. See locate_template().
+ * @param string $type      Sanitized filename without extension.
+ * @param array  $templates A list of template candidates, in descending order of priority.
+ * @return string Empty string to ensure template file is considered not found.
+ */
+function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
+	global $_wp_current_template_hierarchy;
+
+	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
+		$_wp_current_template_hierarchy = $templates;
+	} else {
+		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
+	}
+
+	return '';
+}
+
+/**
+ * Find the correct 'wp_template' post for the current hierarchy and return the path
+ * to the canvas file that will render it.
+ *
+ * @param string $template_file Original template file. Will be overridden.
+ * @return string Path to the canvas file to include.
+ */
+function gutenberg_find_template( $template_file ) {
```

Comment:
Shouldn't we still load a `template file` when all available `wp_template`s have less preference in the hierarchy?

That was one of the goals of #4659.

Reply:
I'm certainly open to it, but let's discuss first. It leads to the question whether we would want to enable the usage of `wp_template` blocks in an "all-or-nothing" mode (e.g. only if something like `add_theme_support( 'block-templates' )` is given) or whether we want to go for a hybrid. The latter at first glance certainly sounds nicer, but it might open a can of worms, as I think supporting both "old" PHP file templates and block templates would make theme development overly complex. Thoughts on this @youknowriad?

Reply:
I see the latter approach as a way for theme authors to opt-in iteratively into block-templates. For example a first step could be to adopt it for "static page templates" only...

Even in terms of shipping strategy, I feel like we can decide to only allow block templates in static ones first and then open to the whole hierarchy as a second step. It's good to start thinking about these discussions but I'm not too concerned at the moment, this will come in time as we move forward, I feel we have room to adapt our approach as we move forward with the implementation.

Reply:
I think the hybrid approach is important, because it allows users to still leverage `wp_template`s in old themes that will never be updated.